service/backend/pchain/block.go

```diff
@@ -2,15 +2,339 @@ package pchain
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/utils/formatting"
+	"github.com/ava-labs/avalanchego/vms/components/avax"
+	"github.com/ava-labs/avalanchego/vms/platformvm"
 	"github.com/coinbase/rosetta-sdk-go/types"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/ava-labs/avalanche-rosetta/mapper"
+	pmapper "github.com/ava-labs/avalanche-rosetta/mapper/pchain"
+	"github.com/ava-labs/avalanche-rosetta/service"
+	"github.com/ava-labs/avalanche-rosetta/service/backend/common"
+	"github.com/ava-labs/avalanche-rosetta/service/backend/pchain/indexer"
+)
+
+var (
+	errMissingBlockIndexHash = errors.New("a positive block index, a block hash or both must be specified")
+	errMismatchedHeight      = errors.New("provided block height does not match height of the block with given hash")
+	errTxInitialize          = errors.New("tx initialize error")
 )
 
+// Block implements the /block endpoint
 func (b *Backend) Block(ctx context.Context, request *types.BlockRequest) (*types.BlockResponse, *types.Error) {
-	return nil, service.ErrNotImplemented
+	isGenesisBlockRequest, err := b.isGenesisBlockRequest(ctx, request.BlockIdentifier)
+	if err != nil {
+		return nil, service.WrapError(service.ErrClientError, err)
+	}
+
+	if isGenesisBlockRequest {
+		block, err := b.buildGenesisBlockResponse(ctx, request.NetworkIdentifier)
+		if err != nil {
+			return nil, service.WrapError(service.ErrClientError, err)
+		}
+		return block, nil
+	}
+
+	var blockIndex int64
+	var hash string
+
+	if request.BlockIdentifier.Index != nil {
+		blockIndex = *request.BlockIdentifier.Index
+	}
+
+	if request.BlockIdentifier.Hash != nil {
+		hash = *request.BlockIdentifier.Hash
+	}
+
+	block, err := b.getBlockDetails(ctx, blockIndex, hash)
+	if err != nil {
+		return nil, service.WrapError(service.ErrClientError, err)
+	}
+	blockIndex = int64(block.Height)
+
+	transactions, err := b.parseTransactions(ctx, request.NetworkIdentifier, block.Txs)
+	if err != nil {
+		return nil, service.WrapError(service.ErrInternalError, err)
+	}
+
+	resp := &types.BlockResponse{
+		Block: &types.Block{
+			BlockIdentifier: &types.BlockIdentifier{
+				Index: blockIndex,
+				Hash:  block.BlockID.String(),
+			},
+			ParentBlockIdentifier: &types.BlockIdentifier{
+				Index: blockIndex - 1,
+				Hash:  block.ParentID.String(),
+			},
+			Timestamp:    block.Timestamp,
+			Transactions: transactions,
+		},
+	}
+
+	return resp, nil
+}
+
+func (b *Backend) buildGenesisBlockResponse(ctx context.Context, networkIdentifier *types.NetworkIdentifier) (*types.BlockResponse, error) {
+	genesisBlock, err := b.getGenesisBlock(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	transactions, err := b.parseTransactions(ctx, networkIdentifier, genesisBlock.Txs)
+	if err != nil {
+		return nil, err
+	}
+
+	return &types.BlockResponse{
+		Block: &types.Block{
+			BlockIdentifier:       b.genesisBlockIdentifier,
+			ParentBlockIdentifier: b.genesisBlockIdentifier,
```

Comment:
We are sticking with the current setup. I have added the following comment explaining the rationale:

```
// Parent block identifier of genesis block is set to itself instead of the hash of the genesis state
// This is done as the genesis state hash cannot be used as a transaction id for the /block apis
// and the operations found in the genesis state are returned as operations of the genesis block.
```

Reply:
LGTM! Thanks.